File: Components/DevExpressVCL/Demos/ExpressVerticalGrid/Delphi/RTTIInspectorDemo/RTTIInspectorDemoVGEditor.pas
unit RTTIInspectorDemoVGEditor;

interface

uses
  Classes, Graphics, Controls, Forms,  Dialogs, cxVGrid, StdCtrls,
  ComCtrls, ExtCtrls, cxLookAndFeelPainters, cxButtons, cxDBVGrid, Menus;

type
  TcxVerticalGridEditor = class(TForm)
    Panel1: TPanel;
    Panel2: TPanel;
    lbRows: TListBox;
    btCategory: TcxButton;
    btEditor: TcxButton;
    btClose: TcxButton;
    btMultiEditor: TcxButton;
    btDelete: TcxButton;
    btClear: TcxButton;
    PopupMenu: TPopupMenu;
    miEditor: TMenuItem;
    miCategory: TMenuItem;
    miMultieditor: TMenuItem;
    N1: TMenuItem;
    miDelete: TMenuItem;
    miClearAll: TMenuItem;
    ShowVerticalGridLayoutEditor: TcxButton;
    procedure btCloseClick(Sender: TObject);
    procedure lbRowsClick(Sender: TObject);
    procedure btCategoryClick(Sender: TObject);
    procedure btEditorClick(Sender: TObject);
    procedure btMultiEditorClick(Sender: TObject);
    procedure btDeleteClick(Sender: TObject);
    procedure btClearClick(Sender: TObject);
    procedure FormActivate(Sender: TObject);
    procedure miEditorClick(Sender: TObject);
    procedure miCategoryClick(Sender: TObject);
    procedure miMultieditorClick(Sender: TObject);
    procedure miDeleteClick(Sender: TObject);
    procedure miClearAllClick(Sender: TObject);
    procedure FormCloseQuery(Sender: TObject; var CanClose: Boolean);
    procedure ShowVerticalGridLayoutEditorClick(Sender: TObject);
    procedure FormClose(Sender: TObject; var Action: TCloseAction);
  private
    FVerticalGrid: TcxCustomVerticalGrid;
    FOnObjectSelected: TNotifyEvent;
    procedure AddRow(ARowClass: TcxCustomRowClass);
    function GetVerticalGrid: TcxCustomVerticalGrid;
    procedure UpdateButtons;
    procedure UpdateItems;
    procedure SetVerticalGrid(const Value: TcxCustomVerticalGrid);
    procedure SelectItem(AItem: Pointer);
  protected
    procedure InitFormEditor;
    procedure DoObjectSelected(Sender: TObject); virtual;
  public
    property OnObjectSelected: TNotifyEvent read FOnObjectSelected write FOnObjectSelected;
    procedure DoItemsModified;
    property VerticalGrid: TcxCustomVerticalGrid read GetVerticalGrid write SetVerticalGrid;
  end;

function GetVerticalGridEditor(AVerticalGrid: TcxCustomVerticalGrid; AEvent: TNotifyEvent): TcxVerticalGridEditor;

implementation

{$R *.dfm}

uses SysUtils;

type
  TcxCustomVerticalGridAccess = class(TcxCustomVerticalGrid);
  TcxVerticalGridCustomizingAccess = class(TcxVerticalGridCustomizing);

const
  SSubStr = 'Tcx';

var
  VerticalGridEditor: TcxVerticalGridEditor;

function GetVerticalGridEditor(AVerticalGrid: TcxCustomVerticalGrid; AEvent: TNotifyEvent): TcxVerticalGridEditor;
begin
  if VerticalGridEditor = nil then
  begin
    VerticalGridEditor := TcxVerticalGridEditor.Create(Application);
    with VerticalGridEditor do
      try
        VerticalGrid := AVerticalGrid;
        OnObjectSelected := AEvent;
        InitFormEditor;
      except
        Free;
      end;
  end;
   Result := VerticalGridEditor;
end;

function LockListBox(AListBox: TListBox): TNotifyEvent;
begin
  Result := AListBox.OnClick;
  AListBox.OnClick := nil;
end;

procedure UnlockListBox(AListBox: TListBox; APrevOnClick: TNotifyEvent);
begin
  AListBox.OnClick := APrevOnClick;
end;

procedure GenVerticalGridRowName(AVerticalGrid: TcxCustomVerticalGrid;
  ARow: TcxCustomRow);
var
  I: Integer;
  S, S1: string;
  ARowName: String;
  function IsComponentNameExist(ANewname: String): Boolean;
  var
    i: Integer;
  begin
    Result := True;
    for i:=0 to AVerticalGrid.Parent.Parent.ComponentCount - 1 do
      if CompareText(AVerticalGrid.Parent.Parent.Components[i].Name, ANewname) = 0 then
        Exit;
    Result := False;
  end;
begin
  S1 := ARow.ClassName;
  I := Pos(SSubStr, S1);
  if I <> 0 then
    S := Copy(S1, I + Length(SSubStr), Length(S1));
  I := 0;
  repeat
    Inc(I);
    ARowName := AVerticalGrid.Name + S + IntToStr(I);
  until not IsComponentNameExist(ARowName);
  ARow.Name := ARowname;
end;

procedure TcxVerticalGridEditor.btCategoryClick(Sender: TObject);
begin
  AddRow(TcxCategoryRow);
end;

procedure TcxVerticalGridEditor.btEditorClick(Sender: TObject);
var
  AIntf: IcxVGridDesignerRows;
begin
  if VerticalGrid.GetInterface(IcxVGridDesignerRows, AIntf) then
    AddRow(AIntf.GetEditorRowClass);
end;

procedure TcxVerticalGridEditor.btMultiEditorClick(Sender: TObject);
var
  AIntf: IcxVGridDesignerRows;
begin
  if VerticalGrid.GetInterface(IcxVGridDesignerRows, AIntf) then
    AddRow(AIntf.GetMultiEditorRowClass);
end;

procedure TcxVerticalGridEditor.btDeleteClick(Sender: TObject);

  function FindItemToSelect: Pointer;
  var
    I: Integer;
  begin
    Result := nil;
    with lbRows do
    begin
      if ItemIndex = -1 then Exit;
      if not Selected[ItemIndex] then
        Result := Items.Objects[ItemIndex]
      else
      begin
        for I := ItemIndex + 1 to Items.Count - 1 do
          if not Selected[I] then
          begin
            Result := Items.Objects[I];
            Exit
          end;
        for I := ItemIndex - 1 downto 0 do
          if not Selected[I] then
          begin
            Result := Items.Objects[I];
            Exit
          end;
      end;
    end;
  end;

var
  AItem: Pointer;
begin
  if lbRows.ItemIndex <> -1 then
  begin
    DoObjectSelected(nil);
    AItem := FindItemToSelect;
    VerticalGrid.Remove(TcxCustomRow(lbRows.Items.Objects[lbRows.ItemIndex]));
    UpdateItems;
    SelectItem(AItem);
    lbRowsClick(nil);
  end;
end;

procedure TcxVerticalGridEditor.btCloseClick(Sender: TObject);
begin
  Close;
end;

procedure TcxVerticalGridEditor.lbRowsClick(Sender: TObject);
begin
  if lbRows.ItemIndex <> -1 then
    DoObjectSelected(lbRows.Items.Objects[lbRows.ItemIndex]);
  UpdateButtons;
end;

procedure TcxVerticalGridEditor.AddRow(ARowClass: TcxCustomRowClass);
var
  ARow: TcxCustomRow;
begin
  if ARowClass <> nil then
  begin
    ARow := VerticalGrid.Add(ARowClass);
    GenVerticalGridRowName(VerticalGrid, ARow);
    UpdateItems;
    SelectItem(ARow);
    UpdateButtons;
  end;
end;

function TcxVerticalGridEditor.GetVerticalGrid: TcxCustomVerticalGrid;
begin
  Result := FVerticalGrid;
end;

procedure TcxVerticalGridEditor.UpdateButtons;
begin
  btDelete.Enabled := lbRows.ItemIndex <> -1;
  miDelete.Enabled := btDelete.Enabled;
  btClear.Enabled := lbRows.Items.Count > 0;
  miClearAll.Enabled := btClear.Enabled;
end;

procedure TcxVerticalGridEditor.UpdateItems;
var
  I, AItemIndex: Integer;
begin
  AItemIndex := lbRows.ItemIndex;
  try
    lbRows.Items.Clear;
    for I := 0 to VerticalGrid.Rows.Count - 1 do
      lbRows.Items.AddObject(VerticalGrid.Rows[I].Name, VerticalGrid.Rows.Items[I]);
  finally
    if AItemIndex <= lbRows.Items.Count - 1 then
      lbRows.ItemIndex := AItemIndex
    else
      lbRows.ItemIndex := lbRows.Items.Count - 1;
  end;
end;

procedure TcxVerticalGridEditor.InitFormEditor;
begin
  UpdateItems;
  UpdateButtons;
end;

procedure TcxVerticalGridEditor.DoItemsModified;
begin
  UpdateItems;
end;

procedure TcxVerticalGridEditor.btClearClick(Sender: TObject);
begin
  DoObjectSelected(nil);
  VerticalGrid.ClearRows;
  UpdateItems;
  UpdateButtons;
end;

procedure TcxVerticalGridEditor.FormActivate(Sender: TObject);
begin
  UpdateButtons;
end;

procedure TcxVerticalGridEditor.miEditorClick(Sender: TObject);
begin
  btEditorClick(nil);
end;

procedure TcxVerticalGridEditor.miCategoryClick(Sender: TObject);
begin
  btCategoryClick(nil);
end;

procedure TcxVerticalGridEditor.miMultieditorClick(Sender: TObject);
begin
  btMultiEditorClick(nil);
end;

procedure TcxVerticalGridEditor.miDeleteClick(Sender: TObject);
begin
  btDeleteClick(nil);
end;

procedure TcxVerticalGridEditor.miClearAllClick(Sender: TObject);
begin
  btClearClick(nil);
end;

procedure TcxVerticalGridEditor.SetVerticalGrid(
  const Value: TcxCustomVerticalGrid);
begin
  FVerticalGrid := Value;
end;

procedure TcxVerticalGridEditor.DoObjectSelected(Sender: TObject);
begin
  if Assigned(FOnObjectSelected) then
    FOnObjectSelected(Sender);
end;

procedure TcxVerticalGridEditor.FormCloseQuery(Sender: TObject;
  var CanClose: Boolean);
begin
  VerticalGridEditor := nil;
end;

procedure TcxVerticalGridEditor.SelectItem(AItem: Pointer);
var
  ItemIndex: Integer;
begin
  ItemIndex := lbRows.Items.IndexOfObject(AItem);
  lbRows.ItemIndex := ItemIndex;
end;

procedure TcxVerticalGridEditor.ShowVerticalGridLayoutEditorClick(
  Sender: TObject);
begin
  with TcxCustomVerticalGridAccess(VerticalGrid) do
  begin
    TcxVerticalGridCustomizingAccess(Customizing).ShowCategoryButtons := True;
    Customizing.Visible := True;
  end;
end;

procedure TcxVerticalGridEditor.FormClose(Sender: TObject;
  var Action: TCloseAction);
begin
  with TcxCustomVerticalGridAccess(VerticalGrid) do
    Customizing.Visible := False;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressVerticalGrid/CBuilder/RTTIInspectorDemo/RTTIInspectorDemoVGEditor.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "RTTIInspectorDemoVGEditor.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxButtons"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxVGrid"
#pragma resource "*.dfm"

class TcxCustomVerticalGridAccess : public TcxCustomVerticalGrid
{
public:
  __property Customizing;
	virtual TcxCustomRowClass __fastcall GetEditorRowClass(void);
	virtual TcxCustomRowClass __fastcall GetMultiEditorRowClass(void);
};

class TcxVerticalGridCustomizingAccess : public TcxVerticalGridCustomizing
{
public:
  __property ShowCategoryButtons;
};

const String SSubStr = "Tcx";

TcxVerticalGridEditor *VerticalGridEditor;

TcxVerticalGridEditor *GetVerticalGridEditor(TcxCustomVerticalGrid *AVerticalGrid,
  TNotifyEvent AEvent)
{
  if (VerticalGridEditor == NULL){
    VerticalGridEditor = new TcxVerticalGridEditor(Application);
    try{
      VerticalGridEditor->VerticalGrid = AVerticalGrid;
      VerticalGridEditor->OnObjectSelected = AEvent;
      VerticalGridEditor->InitFormEditor();
    }
    catch(...){
      delete VerticalGridEditor;
      VerticalGridEditor = NULL;
    }
  }
  return VerticalGridEditor;
}

//---------------------------------------------------------------------------

bool IsComponentNameExist(TcxCustomVerticalGrid *AVerticalGrid, String ANewName)
{
  for (int I = 0; I < AVerticalGrid->Parent->Parent->ComponentCount; I++)
    if (CompareText(AVerticalGrid->Parent->Parent->Components[I]->Name, ANewName) == 0)
      return true;
  return false;
}

void GenVerticalGridRowName(TcxCustomVerticalGrid *AVerticalGrid,
  TcxCustomRow *ARow)
{
  String S1 = ARow->ClassName();
  String S, ARowName;
  int I = S1.Pos(SSubStr);
  if (I != 0)
    S = S1.SubString(I + SSubStr.Length(), S1.Length());
  I = 0;
  do{
    I++;
    ARowName = AVerticalGrid->Name + S + IntToStr(I);
  }while(IsComponentNameExist(AVerticalGrid, ARowName));
  ARow->Name = ARowName;
}

//---------------------------------------------------------------------------
__fastcall TcxVerticalGridEditor::TcxVerticalGridEditor(TComponent* Owner)
  : TForm(Owner)
{
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::btCloseClick(TObject *Sender)
{
  Close();
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::lbRowsClick(TObject *Sender)
{
  if (lbRows->ItemIndex != -1)
    DoObjectSelected(lbRows->Items->Objects[lbRows->ItemIndex]);
  UpdateButtons();
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::btCategoryClick(TObject *Sender)
{
  AddRow(__classid(TcxCategoryRow));
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::btEditorClick(TObject *Sender)
{
  AddRow(((TcxCustomVerticalGridAccess*)VerticalGrid)->GetEditorRowClass());
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::btMultiEditorClick(TObject *Sender)
{
  AddRow(((TcxCustomVerticalGridAccess*)VerticalGrid)->GetMultiEditorRowClass());
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::btDeleteClick(TObject *Sender)
{
  if (lbRows->ItemIndex != -1){
    void *AItem = NULL;
    if (lbRows->ItemIndex != -1){
      if (!lbRows->Selected[lbRows->ItemIndex])
        AItem = lbRows->Items->Objects[lbRows->ItemIndex];
      else{
        for (int I = lbRows->ItemIndex + 1; I < lbRows->Items->Count; I++)
          if (!lbRows->Selected[I]){
            AItem = lbRows->Items->Objects[I];
            break;
          }
        for (int I = lbRows->ItemIndex - 1; I >= 0; I--)
          if (!lbRows->Selected[I]){
            AItem = lbRows->Items->Objects[I];
            break;
          }
      }
    }
    DoObjectSelected(NULL);
    delete (TcxCustomRow*)lbRows->Items->Objects[lbRows->ItemIndex];
    UpdateItems();
    SelectItem(AItem);
    lbRowsClick(NULL);
  }
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::btClearClick(TObject *Sender)
{
  DoObjectSelected(NULL);
  VerticalGrid->ClearRows();
  UpdateItems();
  UpdateButtons();
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::FormActivate(TObject *Sender)
{
  UpdateButtons();
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::miEditorClick(TObject *Sender)
{
  btEditorClick(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::miCategoryClick(TObject *Sender)
{
  btCategoryClick(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::miMultieditorClick(TObject *Sender)
{
  btMultiEditorClick(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::miDeleteClick(TObject *Sender)
{
  btDeleteClick(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::miClearAllClick(TObject *Sender)
{
  btClearClick(NULL);
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::FormCloseQuery(TObject *Sender, bool &CanClose)
{
  VerticalGridEditor = NULL;
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::ShowVerticalGridLayoutEditorClick(TObject *Sender)
{
  TcxCustomVerticalGridAccess *AVGrid = (TcxCustomVerticalGridAccess*)VerticalGrid;
  ((TcxVerticalGridCustomizingAccess*)AVGrid->Customizing)->ShowCategoryButtons = true;
  AVGrid->Customizing->Visible = true;
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::FormClose(TObject *Sender, TCloseAction &Action)
{
  ((TcxCustomVerticalGridAccess*)VerticalGrid)->Customizing->Visible = false;
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::AddRow(TMetaClass* ARowClass)
{
  if (ARowClass != NULL){
    TcxCustomRow *ARow = VerticalGrid->Add(ARowClass);
    GenVerticalGridRowName(VerticalGrid, ARow);
    UpdateItems();
    SelectItem(ARow);
    UpdateButtons();
  }
}
//---------------------------------------------------------------------------

TcxCustomVerticalGrid* __fastcall TcxVerticalGridEditor::GetVerticalGrid(void)
{
  return FVerticalGrid;
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::UpdateButtons()
{
  btDelete->Enabled = lbRows->ItemIndex != -1;
  miDelete->Enabled = btDelete->Enabled;
  btClear->Enabled = lbRows->Items->Count > 0;
  miClearAll->Enabled = btClear->Enabled;
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::UpdateItems()
{
  int AItemIndex = lbRows->ItemIndex;
  try{
    lbRows->Items->Clear();
    for (int I = 0; I < VerticalGrid->Rows->Count; I++)
      lbRows->Items->AddObject(VerticalGrid->Rows->Items[I]->Name,
        VerticalGrid->Rows->Items[I]);
  }
  __finally{
    if (AItemIndex <= lbRows->Items->Count - 1)
      lbRows->ItemIndex = AItemIndex;
    else
      lbRows->ItemIndex = lbRows->Items->Count - 1;
  }
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::SetVerticalGrid(TcxCustomVerticalGrid *Value)
{
  FVerticalGrid = Value;
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::SelectItem(void *AItem)
{
  int AItemIndex = lbRows->Items->IndexOfObject((TObject*)AItem);
  lbRows->ItemIndex = AItemIndex;
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::InitFormEditor()
{
  UpdateItems();
  UpdateButtons();
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::DoObjectSelected(TObject *Sender)
{
  if (FOnObjectSelected != NULL)
    FOnObjectSelected(Sender);
}
//---------------------------------------------------------------------------

void __fastcall TcxVerticalGridEditor::DoItemsModified()
{
  UpdateItems();
}
//---------------------------------------------------------------------------


